packages/flutter/lib/src/material/theme_data.dart

```diff
@@ -73,6 +73,29 @@ export 'package:flutter/services.dart' show Brightness;
 // Examples can assume:
 // late BuildContext context;
 
+/// Defines custom adaptations to a [ThemeData] object.
+///
+/// This is typically used for adaptive themes, but this can be used to adapt any
+/// type of objects. For example, [Switch.adaptive] method produces "cupertino"
+/// [Switch]es on macOS/iOS, and M3 [Switch]es on other platforms. To avoid being
+/// affected by a material-style theme data, [Adaptation] can be used to adapt
+/// the theme data.
+class Adaptation<T> {
+  /// Creates an [Adaptation] constructor.
+  const Adaptation();
+
+  /// The adaptation's type.
+  Type get type => T;
+
+  /// Creates adaptive value which has the same type with the default value.
```

Comment:
Typically overridden to return a custom `ThemeData` class instead of the `defaultValue`. 

Adaptive factory constructors that support adaptations - currently only [Switch.adaptive] - look for a [ThemeData.adaptations] member of the expected type when computing their effective default component theme. 

TBD: this would be a good place to include a  link to the example.